cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java

```diff
@@ -152,17 +157,37 @@ public int hashCode() {
 
       // Create the event and index file with the next sequence id
       int fileSequence = StreamUtils.getNextSequenceId(partitionDirectory, fileNamePrefix);
-      Location eventFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
-                                                            fileSequence, StreamFileType.EVENT);
-      Location indexFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
-                                                            fileSequence, StreamFileType.INDEX);
+      final Location eventFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
+                                                                  fileSequence, StreamFileType.EVENT);
+      final Location indexFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
+                                                                  fileSequence, StreamFileType.INDEX);
       // The creation should succeed, as it's expected to only have one process running per fileNamePrefix.
       if (!eventFile.createNew() || !indexFile.createNew()) {
         throw new IOException("Failed to create new file at " + eventFile + " and " + indexFile);
       }
 
       LOG.debug("New stream file created at {}", eventFile);
-      return new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      final StreamDataFileWriter streamDataFileWriter =
+        new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      timer.schedule(new TimerTask() {
```

Comment:
I have some idea how to do it. Basically at the ConcurrentStreamWriter level, let's say we have one thread that runs once every n minute, which is the the smallest partition time of all streams in the eventQueues map (you can on every run, determine what's the smallest and schedule it's next run time).

Then we add a `ping()` method to the `EventQueue`. What it does is to become leader of that queue (see `tryWrite` method) and call `refresh` method on the current `FileWriter` (maybe define an interface for that method so that it won't mess with the FileWriter interface). The implementation of the `TimePartitionedStreamFileWriter` will implement that interface. Upon refresh is being called, it will close the file if the current partition time is passed.